File: staging/src/github.com/kubeedge/mapper-generator/pkg/common/dataconverter.go
/*
Copyright 2020 The KubeEdge Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package common

import (
	"encoding/json"
	"errors"
	"fmt"
	"reflect"
	"strconv"
	"strings"

	"github.com/golang/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/wrapperspb"
)

// Convert string to other types
func Convert(valueType string, value string) (result interface{}, err error) {
	switch valueType {
	case "int":
		return strconv.ParseInt(value, 10, 64)
	case "float":
		return strconv.ParseFloat(value, 32)
	case "double":
		return strconv.ParseFloat(value, 64)
	case "boolean":
		return strconv.ParseBool(value)
	case "string":
		return value, nil
	default:
		return nil, errors.New("Convert failed")
	}
}

// ConvertToString other types to string
func ConvertToString(value interface{}) (string, error) {
	var result string
	if value == nil {
		return result, nil
	}
	switch v := value.(type) {
	case float64:
		ft := v
		result = strconv.FormatFloat(ft, 'f', -1, 64)
	case float32:
		ft := v
		result = strconv.FormatFloat(float64(ft), 'f', -1, 64)
	case int:
		it := v
		result = strconv.Itoa(it)
	case uint:
		it := v
		result = strconv.Itoa(int(it))
	case int8:
		it := v
		result = strconv.Itoa(int(it))
	case uint8:
		it := v
		result = strconv.Itoa(int(it))
	case int16:
		it := v
		result = strconv.Itoa(int(it))
	case uint16:
		it := v
		result = strconv.Itoa(int(it))
	case int32:
		it := v
		result = strconv.Itoa(int(it))
	case uint32:
		it := v
		result = strconv.Itoa(int(it))
	case int64:
		it := v
		result = strconv.FormatInt(it, 10)
	case uint64:
		it := v
		result = strconv.FormatUint(it, 10)
	case string:
		result = v
	case []byte:
		result = string(v)
	default:
		newValue, err := json.Marshal(value)
		if err != nil {
			return "", err
		}
		result = string(newValue)
	}
	return result, nil
}

// DecodeAnyValue Any to interface
func DecodeAnyValue(value *anypb.Any) (interface{}, error) {
	typeURL := value.GetTypeUrl()

	messageTypeName := getMessageTypeName(typeURL)
	if messageTypeName == "" {
		return nil, fmt.Errorf("cant get message type：%s", typeURL)
	}
	if strings.Contains(messageTypeName, "google.protobuf.") {
		switch messageTypeName {
		case "google.protobuf.Int32Value":
			return decodeWrapperValue(value, &wrapperspb.Int32Value{})
		case "google.protobuf.StringValue":
			return decodeWrapperValue(value, &wrapperspb.StringValue{})
		case "google.protobuf.FloatValue":
			return decodeWrapperValue(value, &wrapperspb.FloatValue{})
		case "google.protobuf.BoolValue":
			return decodeWrapperValue(value, &wrapperspb.BoolValue{})
		case "google.protobuf.Int64Value":
			return decodeWrapperValue(value, &wrapperspb.Int64Value{})
		default:
			return nil, fmt.Errorf("unknown type : %s", messageTypeName)
		}
	}
	messageType := proto.MessageType(messageTypeName)
	if messageType == nil {
		return nil, fmt.Errorf("cant get message type：%s", messageTypeName)
	}

	if !reflect.TypeOf((*proto.Message)(nil)).Elem().AssignableTo(messageType) {
		return nil, fmt.Errorf("assiganbleto proto.Message error：%s", messageTypeName)
	}
	message := reflect.New(messageType.Elem()).Interface().(proto.Message)
	if err := proto.Unmarshal(value.Value, message); err != nil {
		return nil, fmt.Errorf("unmarshal value error：%v", err)
	}
	return message, nil
}

// decodeWrapperValue get proto.Message, then convert to interface
func decodeWrapperValue(value *anypb.Any, wrapper proto.Message) (interface{}, error) {
	if err := proto.Unmarshal(value.Value, wrapper); err != nil {
		return nil, fmt.Errorf("decode wrapperValue,proto unmarshal error：%v", err)
	}
	wrapperValue := reflect.ValueOf(wrapper).Elem()
	valueField := wrapperValue.FieldByName("Value")
	if !valueField.IsValid() {
		return nil, fmt.Errorf("cant get wrapperValue")
	}
	return valueField.Interface(), nil
}

// getMessageTypeName get type by parse type url
func getMessageTypeName(typeURL string) string {
	index := len(typeURL) - 1
	for index >= 0 && typeURL[index] != '/' {
		index--
	}
	if index >= 0 && index < len(typeURL)-1 {
		return typeURL[index+1:]
	}
	return ""
}

File: staging/src/github.com/kubeedge/mapper-framework-Java/src/main/java/data/DataConverter.java
package data;

import com.google.gson.Gson;
import com.google.protobuf.*;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class DataConverter {

    public static Object decodeAnyValue(Any any) throws InvalidProtocolBufferException {
        // Decode any value to Object
        String typeUrl = any.getTypeUrl();
        String messageTypeName = getMessageTypeName(typeUrl);
        if (messageTypeName.isEmpty()){
            log.error("Can not get message type: {}",typeUrl);
        }
        if (messageTypeName.contains("google.protobuf.")){
            switch (messageTypeName){
                case "google.protobuf.Int32Value":
//                    return Int32Value.getDefaultInstance().getParserForType().parseFrom(any.getValue());
                    return Int32Value.parseFrom(any.getValue()).getValue();
                case "google.protobuf.StringValue":
                    return StringValue.parseFrom(any.getValue()).getValue();
                case "google.protobuf.FloatValue":
                    return FloatValue.parseFrom(any.getValue()).getValue();
                case "google.protobuf.BoolValue":
                    return BoolValue.parseFrom(any.getValue()).getValue();
                case "google.protobuf.Int64Value":
                    return Int64Value.parseFrom(any.getValue()).getValue();
                default:
                    log.error("Unknown google.protobuf type: {}",messageTypeName);
                    return null;
            }
        }else{
            log.error("Can not get messageType: {}",messageTypeName);
        }

        return null;
    }

    public static String getMessageTypeName(String typeUrl){// typeUrl例如 type.googleapis.com/google.protobuf.Int32Value
        int lastSlash = typeUrl.lastIndexOf('/');
        if (lastSlash != -1 && lastSlash < typeUrl.length() - 1) {
            return typeUrl.substring(lastSlash + 1);
        }
        return "";
    }

    public static Object convert(String valueType, String value) throws Exception {
        switch (valueType){
            case "int": return Long.parseLong(value);
            case "float": return Float.parseFloat(value);
            case "double": return Double.parseDouble(value);
            case "boolean": return Boolean.parseBoolean(value);
            case "string": return value;
            default:
                log.error("Failed to convert value as {}",valueType);
                throw new Exception("Failed to convert value as " + valueType);
        }
    }

    public static String convertToString(Object value){
        if (value == null){
            return "";
        }else if(value instanceof Number || value instanceof Boolean || value instanceof Character){
            return String.valueOf(value);
        }else if (value instanceof String) {
            return  (String) value;
        } else if (value instanceof byte[]) {
            return new String((byte[]) value);
        } else {
            log.warn("Unknown value Type: {}",value.getClass());
            Gson gson = new Gson();
            try {
                return gson.toJson(value);
            } catch (Exception e) {
                log.error("Failed to convert {} value as string", value.getClass(),e);
            }
        }
        return null;
    }
}
